File: src/tools/web-search.ts
import { loadConfig } from "../config/settings.js";

export const definition = {
  type: "function" as const,
  function: {
    name: "web_search",
    description:
      "Search the web for current information. Use when you need up-to-date data, documentation, or answers not available in local files. Returns top results with titles, URLs, and snippets.",
    parameters: {
      type: "object",
      properties: {
        query: {
          type: "string",
          description: "The search query",
        },
      },
      required: ["query"],
    },
  },
};

interface SearchResult {
  title?: string;
  url?: string;
  snippet?: string;
  content?: string;
}

interface SearchResponse {
  organic_results?: SearchResult[];
  results?: SearchResult[];
  related_searches?: string[];
  [key: string]: any;
}

export async function execute(args: { query: string }): Promise<string> {
  const config = loadConfig();
  if (!config.apiKey) {
    return "Error: No API key configured. Run /config to set it.";
  }

  try {
    const res = await fetch("https://api.minimax.io/v1/coding_plan/search", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${config.apiKey}`,
      },
      body: JSON.stringify({ q: args.query }),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      return `Error: Search API returned ${res.status}${text ? ` — ${text.slice(0, 200)}` : ""}`;
    }

    const data: SearchResponse = await res.json();

    // Extract results — the API may use different field names
    const results = data.organic_results || data.results || [];
    const items = results.slice(0, 8);

    if (items.length === 0) {
      return `No results found for "${args.query}".`;
    }

    const formatted = items.map((r, i) => {
      const title = r.title || "Untitled";
      const url = r.url || "";
      const snippet = r.snippet || r.content || "";
      return `${i + 1}. **${title}**\n   ${url}\n   ${snippet}`;
    });

    let output = formatted.join("\n\n");

    if (data.related_searches && data.related_searches.length > 0) {
      output += `\n\nRelated searches: ${data.related_searches.slice(0, 5).join(", ")}`;
    }

    return output;
  } catch (err: any) {
    if (err.code === "ENOTFOUND" || err.cause?.code === "ENOTFOUND") {
      return "Error: No internet connection.";
    }
    return `Error: ${err.message}`;
  }
}

File: src/tools/web_search.rs
use super::ToolExecutionResult;
use reqwest::header::{HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde_json::Value;

pub fn definition() -> Value {
    serde_json::json!({
        "type": "function",
        "function": {
            "name": "web_search",
            "description": "Search the web for current information. Use when you need up-to-date data, documentation, or answers not available in local files. Returns top results with titles, URLs, and snippets.",
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "The search query"
                    }
                },
                "required": ["query"]
            }
        }
    })
}

pub async fn execute(args: Value) -> ToolExecutionResult {
    let query = args
        .get("query")
        .and_then(|v| v.as_str())
        .unwrap_or("");

    if query.is_empty() {
        return ToolExecutionResult::text("Error: No query provided".to_string());
    }

    // Load API key from config
    let config = crate::config::settings::load_config();
    if config.api_key.is_empty() {
        return ToolExecutionResult::text(
            "Error: No API key configured. Run /config to set it.".to_string(),
        );
    }

    let client = reqwest::Client::new();
    let url = "https://api.minimax.io/v1/coding_plan/search";

    let response = match client
        .post(url)
        .header(CONTENT_TYPE, HeaderValue::from_static("application/json"))
        .header(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {}", config.api_key)).unwrap(),
        )
        .json(&serde_json::json!({ "q": query }))
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => {
            if e.is_connect() {
                return ToolExecutionResult::text("Error: No internet connection.".to_string());
            }
            return ToolExecutionResult::text(format!("Error: {}", e));
        }
    };

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        let preview = if text.len() > 200 {
            &text[..200]
        } else {
            &text
        };
        return ToolExecutionResult::text(format!(
            "Error: Search API returned {}{}",
            status,
            if preview.is_empty() {
                String::new()
            } else {
                format!(" — {}", preview)
            }
        ));
    }

    let data: Value = match response.json().await {
        Ok(d) => d,
        Err(e) => return ToolExecutionResult::text(format!("Error parsing response: {}", e)),
    };

    // Extract results — the API may use different field names
    let results = data
        .get("organic_results")
        .or_else(|| data.get("results"))
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    let items: Vec<&Value> = results.iter().take(8).collect();

    if items.is_empty() {
        return ToolExecutionResult::text(format!("No results found for \"{}\".", query));
    }

    let formatted: Vec<String> = items
        .iter()
        .enumerate()
        .map(|(i, r)| {
            let title = r
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("Untitled");
            let url = r.get("url").and_then(|v| v.as_str()).unwrap_or("");
            let snippet = r
                .get("snippet")
                .or_else(|| r.get("content"))
                .and_then(|v| v.as_str())
                .unwrap_or("");
            format!("{}. **{}**\n   {}\n   {}", i + 1, title, url, snippet)
        })
        .collect();

    let mut output = formatted.join("\n\n");

    if let Some(related) = data.get("related_searches").and_then(|v| v.as_array()) {
        let related_strs: Vec<&str> = related
            .iter()
            .take(5)
            .filter_map(|v| v.as_str())
            .collect();
        if !related_strs.is_empty() {
            output.push_str(&format!(
                "\n\nRelated searches: {}",
                related_strs.join(", ")
            ));
        }
    }

    ToolExecutionResult::text(output)
}
